### PR TITLE
Abstract types for varieties

### PR DESCRIPTION
**Pull request by simonbrandhorst**
We fix our terminology.

In a nutshell:
Affine[/Projective]AlgebraicSet :=  affine [projective] scheme which is geometrically reduced and of finite type over a field
Affine[/Projective / Covered]Variety:= (affine/projective/separated) geometrically integral scheme of finite type over a field 

Curves, Surfaces, Threefolds and ToricVarieties are varieties. 

The plane curves should be integrated into this framework.
There should also be enough room for elliptic curves. @fieker 

@afkafkafk13, @wdecker 
Once we migrate AbsProjectiveScheme out of experimental more will follow.

### CODE REVIEW

**Comment by afkafkafk13** [comment]
@simonbrandhorst : I definitely agree with fixing this terminology -- and the type definitions do precisely this. Do you want a review from me?

**Comment by simonbrandhorst** [comment]
Yes please.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (wdecker) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: What is meant here: The variety is (geometrically) integral, but may be reducible?

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: Copy&Paste error: A **projective**, geometrically...

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: ??? geometrically integral is equivalent to geometrically reduced and geometrically irreducible???
What did you really want to state with this comment?

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: This sentence is somehow hard to read. Maybe
"A curve is a geometrically integral scheme of dimension 1 which is of finite type over a field"

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: Mismatch of word order in the typename.

**Review by afkafkafk13 — COMMENTED** [review]
I left you some comments. 

Only one of them is critical (geometrically integral), the others are of the quality of typos.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl:2]: Per the naming conventions outlined in #1981 (not (yet) in the dev docs), `Abs` should always stand for "Absolute", and `Abstract` should not be abbreviated.

I don't mean this as a blocker for this PR, but we need to resolve this and either change this code (could be in a future PR), or adjust that rule.

CC @fieker

**Comment by wdecker** [comment]
One question is, what does it mean on the practical side, that we work with these types. E.g., I might start with a (complicated) construction yielding some scheme which may or may not be geometrically integral. So then I better do not use the restricted type for the result even if I expect the result to be geometrically integral. So what are the typical situations for using these types? And do we then leave it to the user to make sure that his constructed objects have the desired properties?

**Comment by afkafkafk13** [comment]
From my point of view, these types are for non-specialists. Algebraic geometers will know their way around schemes and not be scared by them. But beginners and users of algebraic geometry (as opposed to algebraic geometers) will need some more basic objects to use for their rather restricted purposes. As soon as they outgrow it, there are always schemes.

For these more restricted types, we should provide a optional argument which allows switching tests of the properties on and off. What the argument should default to, is then a question to be discussed.

**Comment by simonbrandhorst** [comment]
The properties required are always tested. The constructors will have a `check` keyword. 
If you disable the checks, then you compute at your own risk. For instance if you ask for a function field, 
this does not make sense if the variety is not integral. And if it is not geometrically integral then it does not make sense to base change it naively to the algebraic closure.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:2]: We have tons of abstract types. The names will grow very long.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: Thanks, it is a leftover of an older version.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: leftover from an old version. Thanks for catching.

Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl]: I deleted the "Note..." .

**Comment by afkafkafk13** [comment]
You are right: check by default and allow bypass of the check for the experienced user.

**Review by afkafkafk13 — APPROVED** [review]

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:2]: So some abbreviation would be desirable for me. But we should rename this consistently in the schemes folder and not mix `Abstract` and `Abs` in the same ecosystem or even in the same file.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Varieties/Types.jl:1]: one nitpick: 
'Varieties' does not belong on the same level as Objects and Morphisms. I understand, why you put them below Schemes (as they belong to the ecosystem), but in @HereAround 's philosophy they would end up either as neighbouring directory of Schemes or as simple separate file in Schemes/Objects.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:2]: > From my point of view, these types are for non-specialists. Algebraic geometers will know their way around schemes and not be scared by them. But beginners and users of algebraic geometry (as opposed to algebraic geometers) will need some more basic objects to use for their rather restricted purposes. As soon as they outgrow it, there are always schemes.
> 
> For these more restricted types, we should provide a optional argument which allows switching tests of the properties on and off. What the argument should default to, is then a question to be discussed.

In some cases I do not want to work with schemes myself and stay in safe world of varieties,
e.g. when I am considering function fields, rational equivalence and Weil divisors. 
Having the varieties **geometrically** integral assures that one can interpret the results more easily after a base change to an algebraically closed field. 
A question for @wdecker we have absolute primary decomposition over `QQ`. Therefore we can check if a variety is geometrically integral over `QQ`. Can we also do this over a finite field? 
Say I am integral but not geometrically integral. Is there a way to detect this (without computing the absolute primary decomposition)? In Oscar? For geometrically reduced this is no problem because (reduced + over a perfect field) => geometrically reduced.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:1]: Since they rely on schemes they have to be included only after the schemes. 
This was the easiest way to do it. But feel free to change it if you care enough.

Comment (afkafkafk13) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:1]: No, I won't change this. I can live with it. I am looking for stuff by git grep anyway.

File: src/AlgebraicGeometry/Schemes/Varieties/Types.jl
Comment (wdecker) [reply src/AlgebraicGeometry/Schemes/Varieties/Types.jl:2]: We currently can check this ONLY over QQ.

**Pull request merged**
